common/namespace/registry.go

```diff
@@ -483,6 +489,149 @@ func (r *registry) getNamespaceByIDLocked(id ID) (*Namespace, error) {
 	return nil, serviceerror.NewNamespaceNotFound(id.String())
 }
 
+// getOrPutNamespace retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespace(name Name) (*Namespace, error) {
+	// check main caches
+	cacheHit, _ := r.getNamespace(name)
+	if cacheHit != nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(name.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByNamePersistence(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+// getOrPutNamespaceByID retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespaceByID(id ID) (*Namespace, error) {
+	// check main caches
+	cacheHit, _ := r.getNamespaceByID(id)
+	if cacheHit != nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(id.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByIDPersistence(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+func (r *registry) getPreviousReadthroughError(handle string) error {
+	r.readthroughLock.RLock()
+	defer r.readthroughLock.RUnlock()
+
+	if prevErr, ok := r.readthroughCache.Get(handle).(error); ok {
+		return prevErr
+	}
+	return nil
+}
+
+func (r *registry) updateCachesSingleNamespace(ns *Namespace) {
+	var stateChangeCallbacks []StateChangeCallbackFn
+
+	r.cacheLock.Lock()
+	oldNS := r.updateIDToNamespaceCache(r.cacheByID, ns.ID(), ns)
+	r.cacheNameToID.Put(ns.Name(), ns.ID())
+	if namespaceStateChanged(oldNS, ns) {
+		stateChangeCallbacks = mapAnyValues(r.stateChangeCallbacks)
+	}
+	r.cacheLock.Unlock()
+
+	// call state change callbacks
+	deleted := ns.State() == enumspb.NAMESPACE_STATE_DELETED
```

Comment:
the "deleted" bool here means "not present in database", not "present with state == deleted". if we got here then it is present, so the second arg should always be false

Reply:
When getting a single namespace from the database, is there any way to tell if it has been deleted? From what I can tell, we would just get a NotFound error in that case

Reply:
Yeah, if you got it from the database then by definition it's not deleted from the database (it may be "deleted" at the application level but that's different for this purpose)